### PR TITLE
Add namespace targeting mode to CRI and kubelet

### LINKED ISSUE #84359: Ephemeral Containers: Enable container namespace targeting
**What would you like to be added**: Targeting individual container namespaces using ephemeral containers

**Why is this needed**: Currently ephemeral containers get the default namespaces for the pod: if `shareProcessNamespace` is set the ephemeral container runs in the shared process namespace, otherwise it gets an isolated process namespace.

Running in the same process namespace as a target container is really useful for debugging, though, and the ephemeral containers KEP contains a provision for this. This issue is to track its implementation.

/sig node

### PR DESCRIPTION
**Pull request by verb**
**What type of PR is this?**
/kind feature
/sig node
/priority important-longterm

**What this PR does / why we need it**: This adds a TARGET NamespaceMode to the CRI and implements namespace targeting for the PID namespace in the kubelet. This is necessary to implement the TargetContainerName field of EphemeralContainer.

**Which issue(s) this PR fixes**:
Fixes #84359

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
The kubelet and the default docker runtime now support running ephemeral containers in the Linux process namespace of a target container. Other container runtimes must implement this feature before it will be available in that runtime.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://git.k8s.io/enhancements/keps/sig-node/20190212-ephemeral-containers.md
```

### CODE REVIEW

**Comment by verb** [comment]
@yujuhong Shall we try to get this into 1.18?

**Comment by Random-Liu** [comment]
We discussed this in sig-node today. One concern brought up is who is going to reap the ephemeral container processes.

Actually the ephemeral container process targeting a container is similar with `exec` process, there is no much difference.

And here is something we found when we were dealing with an internal exec process leakage:

> I highly suspect that this is caused by the kernel behavior change https://github.com/torvalds/linux/commit/c6c70f4455d1eda91065e93cc4f7eddf4499b105.
> * In linux <4.11, the background exec children are reparented to `containerd-shim` after exec exits, and `containerd-shim` reaps all its child processes. So those processes won't be left zombie.
> * In linux >=4.11, the background exec children are reparented to container init process after exec exists, and most container init processes don't reap arbitrary children.
>
> Based on the kernel change, this doesn't seem to be a kernel bug, it is an expected kernel behavior. If that is the case, we have to live with it.
> 
> I talked with Docker people, and proposed some ideas, but 3 of them need users to change their application:
> 1) Don't run background processes in `exec` (liveness, readiness probes);
> 2) Run a simple init process in their container;
> 3) Switch to shared pod pid namespace if it works for their application, because when pid namespace is shared, the pause process will reap all zombies in the pid namespace. https://github.com/kubernetes/kubernetes/tree/master/build/pause

Given so, I believe:
1) We'll have similar issue with the ephemeral containers with linux 4.11+;
2) We already have this issue with exec today.

**Review by yujuhong — COMMENTED** [review]
> 3\. Switch to shared pod pid namespace if it works for their application, because when pid namespace is shared, the pause process will reap all zombies in the pid namespace. https://github.com/kubernetes/kubernetes/tree/master/build/pause

If the target pod uses shared pid namespace, there wouldn't have reaping problem, right? Maybe we should add this in the feature description to warn the users. 

Also, since this is for debugging/troubleshooting, users may still want to do this even with the possibility of leakage.

**Comment by Random-Liu** [comment]
> Also, since this is for debugging/troubleshooting, users may still want to do this even with the possibility of leakage.

Discussed with @yujuhong offline.

Yeah, during the meeting we only brought up the concern about the process reaping thing, but we didn't know the exact impact. If https://github.com/kubernetes/kubernetes/pull/84731#issuecomment-574314457 is the only impact, it seems fine to us, because:
1) Exec already has the same issue today;
2) Being able to debug a bad pod seems more important than leaking a zombie in the bad pod; :P 
3) Users can avoid this problem after knowing it, e.g. share pid namespace, run a init process, cleanup background processes before exiting the debug container.

Maybe a clear document for this case is good enough?

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go [hunk 1]
```diff
@@ -21,10 +21,15 @@ package kuberuntime
 import (
 	"testing"
 
+	"github.com/google/go-cmp/cmp"
 	"github.com/stretchr/testify/assert"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go]: Why the alias? Same in other places too.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 2]
```diff
@@ -216,6 +216,13 @@ enum NamespaceMode {
     // For example, a container with a PID namespace of NODE expects to view
     // all of the processes on the host running the kubelet.
     NODE      = 2;
+    // TARGET targets the namespace of another container. When this is specified,
+    // a target_id must be specified in NamespaceOption and refer to a container
+    // previously created with NamespaceMode CONTAINER. This containers namespace
+    // will be made to match that of container target_id.
+    // For example, a container with a PID namespace of TARGET expects to view
+    // all of the processes that container target_id can view.
+    TARGET    = 3;
```
Comment (yujuhong) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Looks okay to me.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go (hunk 1 above)
Comment (verb) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go]: > v1 "k8s.io/api/core/v1"

This was added by goimports and I didn't notice. TIL this is intentional: https://github.com/golang/go/issues/30051#issuecomment-495354932

**Comment by yujuhong** [comment]
/lgtm
Will approve soon-ish if no one else has any comments

**Comment by verb** [comment]
@yujuhong rebased due to a merge conflict in `pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go`

Also, I sent kubernetes/enhancements#1533 to you to answer the questions that were asked in the sig-node meeting.

**Comment by verb** [comment]
"Something went wrong: failed to acquire k8s binaries"

/retest

**Review by feiskyer — COMMENTED** [review]
/lgtm

**Comment by verb** [comment]
@yujuhong friendly ping, looks ready for approval?

**Pull request merged**